packages/sanes-chrome-extension/src/extension/backgroundscript.ts

```diff
@@ -1,38 +1,71 @@
 /*global chrome*/
-import { MessageToBackground } from './messages';
-import { handleExternalMessage } from './bsMessageHandler/externalHandler';
-import { handleInternalMessage, getSigningServer } from './bsMessageHandler/internalHandler';
-
-// For a better understanding about the message change done visit:
-// https://developer.chrome.com/extensions/messaging#simple
+import { createPersona, getSigningServer } from './background/createPersona';
+import { RequestMeta } from './background/createPersona/requestCallback';
+import { generateErrorResponse } from './background/errorResponseGenerator';
+import { getPersona } from './background/getPersona';
+import { MessageToBackground, MessageToBackgroundAction } from './background/messages';
 
 /**
  * Listener for dispatching extension requests
  */
 chrome.runtime.onMessage.addListener((message: MessageToBackground, sender, sendResponse) => {
-  handleInternalMessage(message, sender)
-    .then(sendResponse)
-    .catch(console.error);
+  console.log(message, sender);
 
-  // return true to keep the sendResponse reference alive, see
-  // https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/runtime/onMessage#Parameters
-  // However, the Promise return type seems to not work.
-  //
-  // > If you want to asynchronously use sendResponse, add return true; to the onMessage event handler.
+  if (sender.id !== chrome.runtime.id) {
+    return 'Sender is not allowed to perform this action';
+  }
+
+  switch (message.action) {
+    case MessageToBackgroundAction.GetPersona:
+      getPersona()
+        .then(sendResponse)
+        .catch(console.error);
+      break;
+    case MessageToBackgroundAction.CreatePersona:
+      createPersona()
+        .then(sendResponse)
+        .catch(console.error);
+      break;
+    default:
+      return 'Unknown action';
+  }
```

Comment:
Why was handleInternalMessage and handleExternalMessage inlined? I mean both ways are fine but I just want to understand the reason for the change

Reply:
What do you mean by inlined? The main reason to have moved the listeners to `/background/handlers/externalHandler.ts` and `/background/handlers/internalHandler.ts` is because you are not able to import a concrete function (which does not have any chrome reference) if in the same file you have chrome references. And, therefore, JEST was complaining.

In the last commit the `backgroundscript.ts` looks like:
```
/*global chrome*/
import { handleExternalMessage } from './background/handlers/externalHandler';
import { internalHandler } from './background/handlers/internalHandler';

/**
 * Listener for dispatching extension requests
 */
chrome.runtime.onMessage.addListener(internalHandler);

/**
 * Listener for dispatching website requests towards the extension
 */
chrome.runtime.onMessageExternal.addListener(handleExternalMessage);
```

I know maybe is not ideal, but was the best solution I found for unit testing without too much mocking the external messages.

Reply:
In software development, "inline" usually means to take the code of a function's body in put it in the caller code instead of calling the function. One example of _inlining add_ is

Old code

```js
function add(a, b) { return a+b }

const x = add(1, 2)
const y = add(3, 4)
```

New code

```js
const x = 1 + 2
const y = 3 + 4
```

Anyways, in the latest code version `internalHandler` and `handleExternalMessage` are not inlined anymore. They were at the point when I commented, so this is resolved.